### DISCUSSION TITLE
Serverless express component incorrectly reports existing function in update state

### DISCUSSION THREAD

**Discussion by Cameron-C-Chapman** [serverless/serverless#10271] [Q&A]
I'm not sure if this is a bug, and if so where the best place to report it is (this repo, components repo, etc).

Within the last few days I am now having an issue where anytime I try to deploy an express component that has been previously deployed, the cli fails to deploy and reports that the function currently has an update in progress. If I remove the service/component, and re-deploy it works as expected. This is all part of an automated deployment process that has been working for months and there have been no real changes to the service or the deployment process.

For example trying to deploy a service serverless reported this:
```
Serverless › The operation cannot be performed at this time. An update is in progress for resource: arn:aws:lambda:us-east-1:***:function:my-webapp-api-dck8pa-function
```

But if I just use aws directly to check that functions status via the aws cli like
```
aws lambda get-function --region us-east-1 --function-name my-webapp-api-dck8pa-function --query 'Configuration.[State, LastUpdateStatus]
```
I get the output
```
[
    "Active",
    "Successful"
]
```

Removing the service before I deploy is a potential workaround, but that's not ideal and there has to be some issue with the status reporting here. I looked through some recent releases of the cli and components framework but nothing stands out as an issue.

We install the latest serverless cli via npm on every deploy, so we're on the latest version.

Any thoughts on what the issue may be or where I should create an issue/bug?

**Comment by Cameron-C-Chapman** [comment]
Looks like the root cause was new aws state functionality. An explanation of the updates and the way to opt out is documented here: https://aws.amazon.com/blogs/compute/coming-soon-expansion-of-aws-lambda-states-to-all-functions/

I was able to get around this issue by using the description field of my lambda function to opt out of the new aws state functionality.

```
component: express
name: my-app-api

inputs:
  description: aws:states:opt-out
```

It would be nice if there was a little more clarity around the state of serverless components. Is this something that is being phased out in favor of the new cloud offering, will the components registry remain open, etc. It's hard to tell how much of an investment to make in leveraging components when the latest versions of them aren't maintained publicly, and the registry isn't really discoverable. It's serverless's prerogative whether they want to maintain the components system or not, and I'm not complaining, I just want to point out that it's not clear what the current state of the components system is.

**Reply by kaamoss** [reply]
This fix also worked on projects I'm working on. It is worth noting if you are using explicitly named functions, you must edit those directly in the AWS console to have the correct `aws:states:opt-out` hack in the description before attempting a new serverless deploy. Do we know how long this workaround will work, it seems like the description workaround is slated for removal next Monday, Dec 6th?

**Reply by pgrzesik** [reply]
As for `components`, I believe a fix was issued this week that addresses these problems in general. Do you still observe issues with `components` related to Lambda states @kaamoss?

**Reply by kaamoss** [reply]
The project I'm experiencing this issue with is a next.js app using the serverless next.js component/plugin. On an attempted push on December 1st the build failed until adding the new lambda states description text hack @pgrzesik .

**Reply by pgrzesik** [reply]
Hello @kaamoss - I'm assuming it's this component https://github.com/serverless-nextjs. It is based on old version of components which has been deprecated a long time ago. I'm not very familiar with that project but I will reach out internally and see if there's anything that could be done to address it.

**Comment by beppojs** [comment]
I am facing the same issue when deploying a single function via `serverless deploy function --function myfunction --stage prod`. The problem does **not** occur using `serverless deploy --stage prod`

Error message when using `serverless deploy function --function myfunction --stage prod`:
```
The operation cannot be performed at this time. An update is in progress for resource: xyz

Your Environment Information ---------------------------
Operating System:          linux
Node Version:              14.16.1
Framework Version:         2.5.0
Plugin Version:            4.6.0
SDK Version:               n/a
Components Version:        3.18.1
```

After setting the `description` in `serverless.yml` to `description: aws:states:opt-out` the deployment works again. Therefore I agree that this issue is connected to the lambda states update you linked above.

However, the suggested solution is only temporary until December 6 2021 as described in the [article](https://aws.amazon.com/de/blogs/compute/coming-soon-expansion-of-aws-lambda-states-to-all-functions/): 
`NOTE: Delay configuration mechanism has no impact after the end of the Delayed Update. [...] The End of Delayed Update date is now also changed to December 6 2021.`

Any ideas to resolve this issue?

**Reply by pgrzesik** [reply]
Hello @beppojs :wave: Does that happen all the time for you? Would you be able to share output of `deploy function` with `SLS_DEBUG=*` set? I see that you're using a bit outdated v2 version of the Framework, could you update to the latest version?

**Reply by beppojs** [reply]
Thanks @pgrzesik. Unfortunately I am stuck with version 2.5.0 at the moment because of other constraints in my CI/CD pipeline. However, the problem happens all the time.

@Cameron-C-Chapman Since you are using the latest severless version in your CI/CD pipeline, can you share the details for which @pgrzesik has ask for?

**Reply by pgrzesik** [reply]
Thanks for update @beppojs - I've managed to reproduce the behavior on latest versions of the Framework but it's already handled by retry mechanism in `deploy function` logic, I'm going to add additional piece of logic that ensures proper state at the end of `deploy function` execution as well. So, unfortunately, in order to avoid that problem you will need to upgrade your Framework version @beppojs

**Reply by Cameron-C-Chapman** [reply]
@pgrzesik Since you were able to reproduce do you still need any details from me? We are on the latest version of the framework.

If you tag me in the pr/release that addresses this when it's ready I can help validate the fix on my end.

@beppojs For what it's worth, we upgraded from v2 to the latest version awhile ago and didn't run into any issues. I know there's always a lot that goes into upgrading a major version but it was pretty smooth in our experience.

**Reply by pgrzesik** [reply]
@Cameron-C-Chapman I don't think I need any extra details at this point. Components are a bit different beast and I've reported it internally - I will let you know when I get a response about Components specifically.

**Reply by Goodged** [reply]
> Thanks for update @beppojs - I've managed to reproduce the behavior on latest versions of the Framework but it's already handled by retry mechanism in `deploy function` logic, I'm going to add additional piece of logic that ensures proper state at the end of `deploy function` execution as well. So, unfortunately, in order to avoid that problem you will need to upgrade your Framework version @beppojs

When you say it's "already handled" do you mean when it logs "Retrying configuration update for function: ..." over and over until it's able to complete the update?  It seems unfortunate that it used to be possible to do instantaneous function updates, and now it's upwards of 2 minutes waiting for it to retry over and over.  I know nothing of the internals of deploying a function (and the new states added to Lambda causing this issue), but maybe there's a better solution?

**Reply by pgrzesik** [reply]
Hello @Goodged 👋 Thanks a lot for your input on that. Does it take 2 minutes for you all the time? I've been observing very minimal overhead in my scenarios. Is there anything specific about your setup? This is unfortunately something that changed on the Lambda side and I'm afraid we don't have much control over it. Maybe we can investigate potential approaches, but it would be good to have a good reproducible example that exhibits such long deploy times (for me it's max 1 retry at the moment).

**Reply by Goodged** [reply]
@pgrzesik 2 minutes might be an exaggeration (it just feels that way), but I just timed it and it was 30 seconds (~30 retries).  It's certainly never 1 retry...I'd barely have looked into it in that case, haha.  I would say mine is pretty consistently this long though.  This includes when deploying to both my DEV and PROD stages, even though DEV shouldn't have anyone accessing the function (in case that could have been a factor).  Mainly my thought was there might be an AWS call that you need to make to indicate an update was about to occur.

What is the "configuration update" that's being attempted in these retries?  The function has been uploaded, and "successfully deployed".  Do you change the file location of the zip on the lambda to complete the "configuration update"?  I've noticed that during these retries, the new version is not in effect until the configuration update is successful.

Framework Core: 2.67.0
Plugin: 5.5.1
SDK: 4.3.0
Components: 3.18.1

**Reply by pgrzesik** [reply]
Configuration update is about updating function configuration such as `memory`, etc, but if there are no changes to be applied then that step is skipped. 

> Do you change the file location of the zip on the lambda to complete the "configuration update"? I've noticed that during these retries, the new version is not in effect until the configuration update is successful.

No, that's handled by `updateFunctionCode` which then is performed on AWS side - until that update is processed on their side, only then the new version of the function is available, also until that happens, the configuration also cannot be updated, that's why we have the retries for configuration update. So even if we skipped the retries or the configuration update altogether and returned from the command, the new version of the function might still not be accessible.

Is there anything specific about your function configuration?

**Reply by Goodged** [reply]
I ran my deployment using SLS_DEBUG=* and caught the following lines (modified for reasons),


```
Serverless: Successfully deployed function: v1
Serverless: [AWS lambda 409 0.063s 0 retries] updateFunctionConfiguration({
  FunctionName: 'myProject-dev-v1',
  VpcConfig: {
    SecurityGroupIds: [ 'sg-xxxxxxxxxxxxxxxxxxx', 'sg-xxxxxxxxxxxxxxxxxxx', [length]: 2 ],
    SubnetIds: [
      'subnet-xxxxxxxxxxxxxxxxxxx',
      'subnet-xxxxxxxxxxxxxxxxxxx',
      'subnet-xxxxxxxxxxxxxxxxxxx',
      'subnet-xxxxxxxxxxxxxxxxxxx',
      [length]: 4
    ]
  }
})
Serverless: Retrying configuration update for function......
```

So it looks like my yml is confusing the "is an update to the configuration needed".  I have the following defined in the VPC section,

```
provider:
    name: aws
    vpc: ${file(serverless-${self:provider.stage}.yml):vpc}
    ....
```

Which I use to define different security groups and subnets in different stages.  Is this not valid, or should I have done it another way?  Although this doesn't "fix" the retry problem, maybe I shouldn't have hit that section at all?

**Reply by pgrzesik** [reply]
Thanks for the detailed report @Goodged 👍 Indeed, it seems like the logic that checks it doesn't resolve VPC correctly and needs to be improved. 

However, I don't think it's going to help out with waiting for Lambda to reach updated state, because even if configuration update will be skipped, we will need to wait for the function to reach Active State.

**Reply by Goodged** [reply]
Ahh, so you're saying ```updateFunctionCode``` (which replaces the zip) is now putting the function in a non-ready state regardless of the configuration update.  I just read over this URL (https://aws.amazon.com/blogs/compute/tracking-the-state-of-lambda-functions/) which details this life cycle.  It's too bad we can't permanently opt out of the new state system, since it doesn't provide any benefit currently and is much slower (instant vs 30 seconds) than before.

**Reply by pgrzesik** [reply]
That seems to be the case, yes. I've noticed that you're using Lambda inside VPC - I'm guessing that might be causing prolonged lambda state updates but I need to verify, it would be a bummer if it's consistently taking so much time to update 😬

**Reply by Goodged** [reply]
Hmm, interesting theory.  It was the case that an instance of a lambda would take forever to be initialized (6+ seconds) if it was in a VPC...but they fixed that sometime in 2020.  I'm scared to think of what they've done that would have combined this into that :-o

**Reply by pgrzesik** [reply]
Hey @Goodged - I've managed to test it out on a Lambda function with VPC and it's much worse - it takes around 30 seconds at least to upgrade a function. If I remove the VPC then it's back to reasonable update time. I'm afraid that it might have to do something with VPCs specifically 😬

**Reply by Goodged** [reply]
Ooof.  Well, at least my issue is completely reproducible.  Hopefully they (AWS) will do some work on that cuz it's pretty painful for development.  I presume in yours, you didn't have it detecting a change in the VPC, and therefore try to do an update that wasn't required like mine was?

**Reply by pgrzesik** [reply]
It was still doing that as it's not yet fixed on v2, but even without it, it required a long wait (30 seconds or something around that) for function to reach `Active` state again.

**Reply by beppojs** [reply]
I managed to update serverless to `2.69.1` now and can confirm that the fix by @pgrzesik in `2.68.0` works for me and resolved the problem described above. Thanks for resolving that issue!

**Reply by pgrzesik** [reply]
Glad to hear it's working fine for you @beppojs 🙌

**Reply by Goodged** [reply]
Not sure what the change was in 2.68.0, but I just updated to 2.69.1 and I'd say things are now worse than before, haha.  Instead of just getting the "Retrying configuration update for function", I now get the following

```
<30 times> Serverless: Retrying configuration update for function: funcName. Reason: The operation cannot be performed at this time. An update is in progress for resource: arn:aws:lambda:us-west-2:XXXXX:funcName
<1 time> Serverless: Successfully updated function: funcName
<43 times> Serverless: Retrying ensure function state for function: funcName.  
```
The "retrying ensure function state" message 43 times is new  :(

**Reply by pgrzesik** [reply]
Hello @Goodged - this is an expected behavior, sadly. From now on, `deploy function` will always ensure that update function actually finished updating and reached active state again, otherwise you would encounter weird behavior where function deployment "finished" from your perspective but if you call the function, still the old version is used. Are you using VPC with that function by any chance?

**Reply by Goodged** [reply]
Thanks for confirming this is expected behavior.  Yes, all my functions are in a VPC which previously sounded like the cause of these long update times now

**Reply by pgrzesik** [reply]
It seems that VPC-based functions are now slower when it comes to updates, if I have any additional info around that I will report it in this thread.
